specs/discussions/2025-04-block-proofs.md

```diff
@@ -0,0 +1,319 @@
+
+/// Enum to represent the validation result of a block submission.
+enum ValidationResult {
+	Valid,
+	ReOrg,
+}
+
+/// Validates the submitted block header against the following conditions and determines
+/// whether the block can be appended to the child shard history or if a reorg is required.
+fn validate_block_submission(block_header: &BlockHeader, parent_chain: &ParentChain) -> Result<ValidationResult, Error> {
+	// 1. Check if the block references a valid beacon chain block.
+	if !beacon_chain.is_valid_block(block_header.beacon_chain_ref) {
+		return Err(Error::InvalidBeaconChainReference);
+	}
+
+	// 2. Verify that the solution is within the acceptable range.
+	if !is_solution_in_range(block_header.solution) {
+		return Err(Error::InvalidSolutionRange);
+	}
+
+	// 3. Ensure the block number is increasing or handle reorg scenarios.
+	let latest_committed_block = parent_chain.get_latest_committed_block(block_header.shard_id);
+	if block_header.number <= latest_committed_block.number {
+		if is_reorg_valid(block_header, latest_committed_block) {
+			return Ok(ValidationResult::ReOrg);
+		} else {
+			return Err(Error::InvalidBlockNumberOrReorg);
+		}
+	} else if block_header.number != latest_committed_block.number + 1 {
+		return Err(Error::NonSequentialBlockNumber);
+	}
+
+	// 4. Verify that the block points to the correct parent block.
+	let expected_parent_hash = latest_committed_block.hash;
+	if block_header.parent_hash != expected_parent_hash {
+		if is_reorg_valid(block_header, latest_committed_block) {
+			return Ok(ValidationResult::ReOrg);
+		} else {
+			return Err(Error::InvalidParentBlockReference);
+		}
+	}
+
+	Ok(ValidationResult::Valid)
+}
+
+/// Helper function to check if a reorg is valid.
+fn is_reorg_valid(block_header: &BlockHeader, latest_committed_block: &BlockHeader) -> bool {
+	// Check if the block number is lower than the latest committed block
+	// and if the parent chain's view of the child shard's history aligns with the reorg.
+	let parent_view = parent_chain.get_child_shard_history(block_header.shard_id);
+	parent_view.is_consistent_with_reorg(block_header)
+}
+
+/// Helper function to check if the solution is within the valid range.
+fn is_solution_in_range(solution: &Proof) -> bool {
+	// Current logic for solution verification
+	// NOTE: (it may change slightly in the future once we define better how sharded farming will work)
+	true
+}
+```
+
+4. If the verification is successful, the following information is stored in the parent shard for
+   each shard. This information can be used to generate inclusion proofs for the blocks in the child
+   shard and compare the with the information submitted in the parent chain.
+
+```rust
+/// Data structure that is kept on-chain with the hash of the block submitted from the child shard.
+let shardBlocksMap = HashMap<ShardId, HashMap<BlockNumber, Hash>>
+```
+
+5. Full nodes in child shards represent their own history as Mountain Merkle Roots (MMRs) whose root
```

Comment:
The point of MMR is to be able to prove anything that happened at any point in the past. I'm not sure what this has to do with global (I assume archival) history. Maybe you can elaborate?

Reply:
Totally. What I was trying to describe here is that the MRR allows the generation of proofs of the history of the child shard, but by committing blocks in the parent chain (and recursively in the beacon chain), there is a way to link this history proofs back to the history of the beacon chain.

Reply:
I think what confused be a bit is this part:
> without having to wait for their segments to be included in the global history in the beacon chain

It wasn't clear how MMRs can help bypass the requirement of the shard being included in the beacon chain. To me it looks complementary, not a replacement.

Reply:
I see. Rephrased it a bit to make a bit clear what I mean.